whetstone.py

```diff
@@ -39,15 +61,304 @@ def _encode_credentials(self):
         encoded_credentials_string = str(encoded_credentials, "utf-8")
         return "Basic " + encoded_credentials_string
 
-    def get_all(self, endpoint, tag=False):
-        if tag:
-            endpoint_url = f"{self.url}/external/generic-tags/{endpoint}"
+    def get_all(self):
+        if self.tag:
+            endpoint_url = f"{self.url}/external/generic-tags/{self.endpoint}"
         else:
-            endpoint_url = f"{self.url}/external/{endpoint}"
+            endpoint_url = f"{self.url}/external/{self.endpoint}"
         headers = {"Authorization": f"Bearer {self.token}"}
         response = requests.get(endpoint_url, headers=headers)
         if response.status_code == 200:
             response_json = response.json()
             return response_json["data"]
         else:
-            raise Exception(f"Failed to list {endpoint}")
+            raise Exception(f"Failed to list {self.endpoint}")
+
+    def _write_to_db(self, df):
+        self.sql.insert_into(self.table_name, df, chunksize=10000, if_exists="replace")
+
+    def _write_to_json(self, data):
+        if os.path.exists(self.filename):
+            os.remove(self.filename)
+        with open(self.filename, "w") as f:
+            json.dump(data, f, indent=2)
+
+    def import_all(self):
+        data = self.get_all()
+        self._write_to_json(data)
+        self.additional_imports(data)
+        df = self._process_and_filter_records(data)
+        self._write_to_db(df)
+
+    def _preprocess_records(self, records):
+        return records
+
+    def _process_and_filter_records(self, records):
+        new_records = self._preprocess_records(records)
+        df = pd.json_normalize(new_records)
+        df = df.reindex(columns=self.columns)
+        df = df.astype("object")
+        df = self._convert_dates(df)
+        df.rename(columns={"_id": "id"}, inplace=True)
+        return df
+
+    def _convert_dates(self, df):
+        dates = [col for col in df.columns if col in self.dates]
+        if dates:
+            date_types = {col: "datetime64[ns]" for col in dates}
+            df = df.astype(date_types)
+        return df
```

Comment:
@zkagin This works a little differently. I've defined the list of all possible date columns in the parent class. Within child classes, this checks for the existence of those columns in the request object and then converts the ones in the request that match.

Reply:
Oh I see now, it is first finding columns that are also in the global date list, and then creating the dictionary from that.

I think this might work then?
```
date_types = {col: "datetime64[ns]" for col in df.columns if col in self.dates}
if date_types:
    df = df.astype(date_types)
return df
```